Flaky InstallRequestHandlingTest.shouldBeAbleToBecomeLeader

**Summary**

> Try to answer the following as best as possible

- How often does the test fail? 1
- Does it block your work? No
- Do we suspect that it is a real failure? no

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
java.util.concurrent.CompletionException: java.net.ConnectException: Expected to send a message with subject 'raft-partition-partition-1-transfer' to member '0', but member is not known. Known members are '[Member{id=1, address=0.0.0.0:2146, properties={brokerInfo=EADJAAAAAwABAAAAAQAAAAMAAAADAAAAAAABCgAAAGNvbW1hbmRBcGkMAAAAMC4wLjAuMDoyMTQ1BQABAQAAAAEMAAAOAAAAOC4xLjAtU05BUFNIT1QFAAEBAAAAAQ==}}, Member{id=2, address=0.0.0.0:2150, properties={brokerInfo=EADJAAAAAwACAAAAAQAAAAMAAAADAAAAAAABCgAAAGNvbW1hbmRBcGkMAAAAMC4wLjAuMDoyMTQ5BQABAQAAAAEMAAAOAAAAOC4xLjAtU05BUFNIT1QFAAEBAAAAAQ==}}, Member{id=gateway, address=0.0.0.0:2152, properties={event-service-topics-subscribed=KIIDAGpvYnNBdmFpbGFibOU=}}]'.
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
	at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:674)
	at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:662)
	at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2168)
	at io.atomix.cluster.messaging.impl.DefaultClusterCommunicationService.send(DefaultClusterCommunicationService.java:115)
	at io.atomix.raft.partition.impl.RaftServerCommunicator.sendAndReceive(RaftServerCommunicator.java:216)
	at io.atomix.raft.partition.impl.RaftServerCommunicator.transfer(RaftServerCommunicator.java:89)
	at io.atomix.raft.impl.RaftContext.lambda$anoint$33(RaftContext.java:563)
	at io.atomix.utils.concurrent.SingleThreadContext$WrappedRunnable.run(SingleThreadContext.java:171)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.net.ConnectException: Expected to send a message with subject 'raft-partition-partition-1-transfer' to member '0', but member is not known. Known members are '[Member{id=1, address=0.0.0.0:2146, properties={brokerInfo=EADJAAAAAwABAAAAAQAAAAMAAAADAAAAAAABCgAAAGNvbW1hbmRBcGkMAAAAMC4wLjAuMDoyMTQ1BQABAQAAAAEMAAAOAAAAOC4xLjAtU05BUFNIT1QFAAEBAAAAAQ==}}, Member{id=2, address=0.0.0.0:2150, properties={brokerInfo=EADJAAAAAwACAAAAAQAAAAMAAAADAAAAAAABCgAAAGNvbW1hbmRBcGkMAAAAMC4wLjAuMDoyMTQ5BQABAQAAAAEMAAAOAAAAOC4xLjAtU05BUFNIT1QFAAEBAAAAAQ==}}, Member{id=gateway, address=0.0.0.0:2152, properties={event-service-topics-subscribed=KIIDAGpvYnNBdmFpbGFibOU=}}]'.
	at io.atomix.cluster.messaging.impl.DefaultClusterCommunicationService.failOnMemberNotKnown(DefaultClusterCommunicationService.java:235)
	at io.atomix.cluster.messaging.impl.DefaultClusterCommunicationService.sendAndReceive(DefaultClusterCommunicationService.java:223)
	at io.atomix.cluster.messaging.impl.DefaultClusterCommunicationService.send(DefaultClusterCommunicationService.java:114)
	... 10 more
</pre>
</details>

**Hypotheses**

None.

**Logs**

[Build](https://github.com/camunda/zeebe/runs/7560828899?check_suite_focus=true)

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/BrokerLeaderChangeTest.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/BrokerLeaderChangeTest.java
@@ -63,7 +63,7 @@ public final class BrokerLeaderChangeTest {
     final var firstLeaderNodeId = firstLeaderInfo.getNodeId();
 
     // when
-    clusteringRule.forceClusterToHaveNewLeader(1, firstLeaderNodeId);
+    clusteringRule.forceNewLeaderForPartition(firstLeaderNodeId, 1);
 
     // then
     assertThat(clusteringRule.getCurrentLeaderForPartition(1).getNodeId())

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/BrokerLeaderChangeTest.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/BrokerLeaderChangeTest.java
@@ -63,7 +63,7 @@ public final class BrokerLeaderChangeTest {
     final var firstLeaderNodeId = firstLeaderInfo.getNodeId();
 
     // when
-    clusteringRule.forceClusterToHaveNewLeader(firstLeaderNodeId);
+    clusteringRule.forceClusterToHaveNewLeader(1, firstLeaderNodeId);
 
     // then
     assertThat(clusteringRule.getCurrentLeaderForPartition(1).getNodeId())

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/ClusteringRule.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/ClusteringRule.java
@@ -15,6 +15,7 @@ import static io.camunda.zeebe.broker.test.EmbeddedBrokerConfigurator.setCluster
 import static io.camunda.zeebe.broker.test.EmbeddedBrokerConfigurator.setInitialContactPoints;
 import static io.camunda.zeebe.broker.test.EmbeddedBrokerRule.assignSocketAddresses;
 import static io.camunda.zeebe.protocol.Protocol.START_PARTITION_ID;
+import static org.assertj.core.api.Assertions.assertThat;
 
 import io.atomix.cluster.AtomixCluster;
 import io.atomix.cluster.AtomixClusterBuilder;
@@ -602,27 +603,38 @@ public final class ClusteringRule extends ExternalResource {
     }
   }
 
-  public void forceClusterToHaveNewLeader(final int expectedLeader) {
-    final var previousLeader = getCurrentLeaderForPartition(1);
-    if (previousLeader.getNodeId() == expectedLeader) {
+  public void forceClusterToHaveNewLeader(final int partitionId, final int expectedLeaderId) {
+    final var previousLeader = getCurrentLeaderForPartition(partitionId);
+    final var expectedLeader = brokers.get(expectedLeaderId);
+
+    if (previousLeader.getNodeId() == expectedLeaderId) {
       return;
     }
 
-    final var broker = brokers.get(expectedLeader);
-    final var atomix = broker.getBrokerContext().getClusterServices();
-    final MemberId nodeId = atomix.getMembershipService().getLocalMember().id();
+    final var serverOfExpectedLeader =
+        ((RaftPartition)
+                expectedLeader
+                    .getBrokerContext()
+                    .getPartitionManager()
+                    .getPartitionGroup()
+                    .getPartition(partitionId))
+            .getServer();
 
-    final var raftPartition =
-        broker.getBrokerContext().getPartitionManager().getPartitionGroup().getPartitions().stream()
-            .filter(partition -> partition.members().contains(nodeId))
-            .filter(partition -> partition.id().id() == START_PARTITION_ID)
-            .map(RaftPartition.class::cast)
-            .findFirst()
-            .orElseThrow();
+    Awaitility.await("Promote request is successful")
+        .pollInterval(Duration.ofMillis(500))
+        .timeout(Duration.ofMinutes(1))
+        .untilAsserted(
+            () ->
+                assertThat(serverOfExpectedLeader.promote())
+                    .succeedsWithin(Duration.ofSeconds(15)));
 
-    raftPartition.getServer().promote().join();
-
-    awaitOtherLeader(START_PARTITION_ID, previousLeader.getNodeId());
+    Awaitility.await("New leader of partition %s is %s".formatted(partitionId, expectedLeaderId))
+        .pollInterval(Duration.ofMillis(500))
+        .atMost(Duration.ofMinutes(1))
+        .ignoreExceptions()
+        .until(
+            () -> getLeaderForPartition(partitionId),
+            (leader) -> leader.getNodeId() == expectedLeaderId);
   }
 
   public void waitForTopology(final Consumer<TopologyAssert> assertions) {

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/ClusteringRule.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/ClusteringRule.java
@@ -603,7 +603,7 @@ public final class ClusteringRule extends ExternalResource {
     }
   }
 
-  public void forceClusterToHaveNewLeader(final int partitionId, final int expectedLeaderId) {
+  public void forceNewLeaderForPartition(final int expectedLeaderId, final int partitionId) {
     final var previousLeader = getCurrentLeaderForPartition(partitionId);
     final var expectedLeader = brokers.get(expectedLeaderId);
 

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/InstallRequestHandlingTest.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/InstallRequestHandlingTest.java
@@ -52,7 +52,7 @@ public class InstallRequestHandlingTest {
 
     // then
     clusteringRule.waitForSnapshotAtBroker(followerId);
-    clusteringRule.forceClusterToHaveNewLeader(followerId);
+    clusteringRule.forceClusterToHaveNewLeader(1, followerId);
 
     clientRule.getClient().newCompleteCommand(jobKey).send().join();
     ZeebeAssertHelper.assertProcessInstanceCompleted("process");
@@ -79,7 +79,7 @@ public class InstallRequestHandlingTest {
     clientRule.createProcessInstance(processDefinitionKey);
 
     // then
-    clusteringRule.forceClusterToHaveNewLeader(followerId);
+    clusteringRule.forceClusterToHaveNewLeader(1, followerId);
 
     final var jobKey =
         RecordingExporter.jobRecords(JobIntent.CREATED)

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/InstallRequestHandlingTest.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/InstallRequestHandlingTest.java
@@ -52,7 +52,7 @@ public class InstallRequestHandlingTest {
 
     // then
     clusteringRule.waitForSnapshotAtBroker(followerId);
-    clusteringRule.forceClusterToHaveNewLeader(1, followerId);
+    clusteringRule.forceNewLeaderForPartition(followerId, 1);
 
     clientRule.getClient().newCompleteCommand(jobKey).send().join();
     ZeebeAssertHelper.assertProcessInstanceCompleted("process");
@@ -79,7 +79,7 @@ public class InstallRequestHandlingTest {
     clientRule.createProcessInstance(processDefinitionKey);
 
     // then
-    clusteringRule.forceClusterToHaveNewLeader(1, followerId);
+    clusteringRule.forceNewLeaderForPartition(followerId, 1);
 
     final var jobKey =
         RecordingExporter.jobRecords(JobIntent.CREATED)

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/ReaderCloseTest.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/ReaderCloseTest.java
@@ -68,7 +68,7 @@ public class ReaderCloseTest {
             .getConfig()
             .getCluster()
             .getNodeId();
-    clusteringRule.forceClusterToHaveNewLeader(1, followerId);
+    clusteringRule.forceNewLeaderForPartition(followerId, 1);
     // because of https://github.com/camunda/zeebe/issues/8329
     // we need to add another record so we can do a snapshot
     clientRule

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/ReaderCloseTest.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/ReaderCloseTest.java
@@ -68,7 +68,7 @@ public class ReaderCloseTest {
             .getConfig()
             .getCluster()
             .getNodeId();
-    clusteringRule.forceClusterToHaveNewLeader(followerId);
+    clusteringRule.forceClusterToHaveNewLeader(1, followerId);
     // because of https://github.com/camunda/zeebe/issues/8329
     // we need to add another record so we can do a snapshot
     clientRule

